ecs-agent/acs/session/task_manifest_responder_test.go

```diff
@@ -0,0 +1,228 @@
+		MessageId:            aws.String(testconst.MessageID),
+		Tasks: []*ecsacs.TaskIdentifier{
+			{DesiredStatus: aws.String("STOPPED"), TaskArn: aws.String(testconst.TaskARN)},
+		},
+		Timeline: aws.Int64(20),
+	}
+
+	// Handle the task manifest message update
+	handleManifestMessage := testTaskManifestResponder.HandlerFunc().(func(*ecsacs.TaskManifestMessage))
+	handleManifestMessage(testManifest)
+
+	ackWg.Wait()
+
+	// Now assert that the latest sequence number has been updated
+
+	// Now manually try to reset the sequence and make sure it does not update
+	// As 15 < 20 we should not be able to update to an older sequence number
+	testTaskManifestResponder.snAccessor.SetLatestSequenceNumber(testManifest)
+
+	require.NotEqual(t,
+		aws.Int64(testTaskManifestResponder.snAccessor.GetLatestSequenceNumber()),
+		aws.Int64(StartingLatestTaskManifestSequenceNumber))
+}
+
+func TestTaskStopVerification(t *testing.T) {
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	stopVerifSent := make(chan *ecsacs.TaskStopVerificationMessage)
+
+	testResponseSender := func(response interface{}) error {
+		ackRequest, isAckRequest := response.(*ecsacs.AckRequest)
+		if isAckRequest {
+			require.Equal(t, aws.StringValue(ackRequest.MessageId), testconst.MessageID)
+		} else {
+			stopVerifSent <- response.(*ecsacs.TaskStopVerificationMessage)
+		}
+		return nil
+	}
+
+	testTaskManifestResponder := setupTestTaskManifestResponder(t, metrics.NewNopEntryFactory(), testResponseSender)
+
+	// After we get the latest sequence number, we will expect to set the latest message ID
+	mockmmIDA := testTaskManifestResponder.messageIDAccessor.(*mock_session.MockManifestMessageIDAccessor)
+	mockmmIDA.EXPECT().SetMessageID(gomock.Any()).Times(1)
+
+	// Expect to get the latest sequence number for stale check, then set it as we expect to pass
+	mockSNA := testTaskManifestResponder.snAccessor.(*mock_session.MockSequenceNumberAccessor)
+	mockSNA.EXPECT().GetLatestSequenceNumber().Times(1)
+	mockSNA.EXPECT().SetLatestSequenceNumber(StartingLatestTaskManifestSequenceNumber).Times(1)
+
+	// Create test task manifest.
+	testManifest := setupTestManifestMessage()
+
+	// Mock the task comparer to return a running task which will spur a task stop verification message
+	mockComparer := testTaskManifestResponder.taskComparer.(*mock_session.MockTaskComparer)
+	mockComparer.EXPECT().CompareRunningTasksOnInstanceWithManifest(gomock.Any()).Return(
+		[]*ecsacs.TaskIdentifier{
+			{DesiredStatus: aws.String("RUNNING"), TaskArn: aws.String(testconst.TaskARN)},
+		},
+		nil,
+	)
+
+	// Handle the task manifest message update
+	handleManifestMessage := testTaskManifestResponder.HandlerFunc().(func(*ecsacs.TaskManifestMessage))
+	go handleManifestMessage(testManifest)
+
+	// Wait for the ack to send through our channel so we can receive it
+	taskStopVerifSent := <-stopVerifSent
+
+	// We expect only one task marked for termination, and it should be running
```

Comment:
If we expect it to be running, should we assert that? Also, I'm assuming we only expect it to be running because we have set the desired status to RUNNING on L212. However FWIW, I'm not sure in the real world if we would ever receive any task here that doesn't have desired status STOPPED. Could we change the expectation of this test to use desired status STOPPED instead to more match what we expect to happen in normal operation?

Reply:
Actually, L212 acts as an injection into the test so we can force the method to return the running task. I think cancelling a running task is exactly the use case for this component, as a stopped or stopping task is already being handled and would render this responder redundant. It is also not up to this responder to change task state or stop the task, that responsibility lies with ACS.

Reply:
> Actually, L212 acts as an injection into the test so we can force the method to return the running task. I think cancelling a running task is exactly the use case for this component, as a stopped or stopping task is already being handled and would render this responder redundant. 

In normal operations we wouldn't expect the method to return a task with desired status RUNNING. Currently, the list of tasks returned by this method as part of this flow always have desired status STOPPED. I've sent some links offline to you to point to this behavior. 

> It is also not up to this responder to change task state or stop the task, that responsibility lies with ACS.

In the context of the full task sync feature, this responder is not responsible to stop the task. However, Agent is the one actually stopping the task but that is done via the task stop verification ACK responder.

Reply:
ack